support/scheduler/schedule.go

```diff
@@ -0,0 +1,442 @@
+		deleteScheduleOperation(scheduleId, scheduleContext)
+	}
+
+	loggingClient.Info("removed the schedule event with id " + scheduleEventId)
+
+	return nil
+}
+
+//endregion
+
+//region schedule event execution
+func triggerSchedule() {
+	nowEpoch := time.Now().Unix()
+
+	defer func() {
+		if err := recover(); err != nil {
+			loggingClient.Error("trigger schedule error : " + err.(string))
+		}
+	}()
+
+	if scheduleQueue.Length() == 0 {
+		return
+	}
+
+	loggingClient.Debug(fmt.Sprintf("%d item in schedule queue.", scheduleQueue.Length()))
+
+	for i := 0; i < scheduleQueue.Length(); i++ {
+		if scheduleQueue.Peek().(*ScheduleContext) != nil {
+			scheduleContext := scheduleQueue.Remove().(*ScheduleContext)
+			scheduleId := scheduleContext.Schedule.Id.Hex()
+			loggingClient.Debug("check schedule with id : " + scheduleId)
+			if scheduleContext.MarkedDeleted {
+				loggingClient.Debug("the schedule with id : " + scheduleId + " be marked as deleted, removing it.")
+				continue //really delete from the queue
+			} else {
+				loggingClient.Debug("schedule with id : " + scheduleId + " next schedule time is : " + scheduleContext.NextTime.String())
+				if scheduleContext.NextTime.Unix() <= nowEpoch {
+					loggingClient.Info("executing schedule, detail : {" + scheduleContext.GetInfo() + "} , at : " + scheduleContext.NextTime.String())
+					execute(scheduleContext)
+
+					scheduleContext.UpdateNextTime()
+					scheduleContext.UpdateIterations()
+
+					if scheduleContext.IsComplete() {
+						loggingClient.Info("completed schedule, detail : " + scheduleContext.GetInfo())
+					} else {
+						loggingClient.Info("requeue schedule, detail : " + scheduleContext.GetInfo())
+						scheduleQueue.Add(scheduleContext)
+					}
+				} else {
+					scheduleQueue.Add(scheduleContext)
+				}
+			}
+		}
+	}
+}
+
+func execute(context *ScheduleContext) {
+	scheduleEventsMap := context.ScheduleEventsMap
+
+	defer func() {
+		if err := recover(); err != nil {
+			loggingClient.Error("schedule execution error : " + err.(string))
+		}
+	}()
+
+	loggingClient.Debug(fmt.Sprintf("%d schedule event need to be executed.", len(scheduleEventsMap)))
+
+	//execute schedule event one by one
+	for eventId := range scheduleEventsMap {
+		loggingClient.Info("the event with id : " + eventId + " belongs schedule : " + context.Schedule.Id.Hex() + " is be executing!")
+		scheduleEvent, _ := scheduleEventsMap[eventId]
+
+		executingUrl := getUrlStr(scheduleEvent.Addressable)
+		loggingClient.Debug("the event with id : " + eventId + " will request url : " + executingUrl)
+
+		req, err := http.NewRequest(http.MethodPost, executingUrl, nil)
+		req.Header.Set(ContentTypeKey, ContentTypeJsonValue)
+		req.Header.Set(ContentLengthKey, string(len(scheduleEvent.Parameters)))
+
+		if err != nil {
+			loggingClient.Error("create new request occurs error : " + err.Error())
+		}
+
+		client := &http.Client{
```

Comment:
The direct use of an http client makes this class impossible to unit test. Also, this logic is never called by the current tests. I verified this using GoLand IDE Debugger.

A better strategy would be to mock any HTTP interaction through a client interface. Examples here:
https://github.com/edgexfoundry-holding/core-data-refactor/blob/master/core/aggregates/events/event.go

Reply:
will consider this later.